crypto/bn/bn_prime.c

```diff
@@ -415,21 +415,18 @@ static int probable_prime(BIGNUM *rnd, int bits, prime_t *mods)
     delta = 0;
  loop:
     if (is_single_word) {
-        BN_ULONG rnd_word = BN_get_word(rnd);
-
-        /*-
-         * In the case that the candidate prime is a single word then
-         * we check that:
-         *   1) It's greater than primes[i] because we shouldn't reject
-         *      3 as being a prime number because it's a multiple of
-         *      three.
-         *   2) That it's not a multiple of a known prime. We don't
-         *      check that rnd-1 is also coprime to all the known
-         *      primes because there aren't many small primes where
-         *      that's true.
+        BN_ULONG rnd_word = BN_get_word(rnd) + delta;
+
+        /*
+         * check that rnd is not a prime and also that gcd(rnd-1,primes)
+         * == 1 (except for 2)
+         * do the second check only is we are interested in safe primes
+         * in the case that the candidate prime is a single word then
+         * we check only the primes up to sqrt(rnd)
          */
-        for (i = 1; i < NUMPRIMES && primes[i] < rnd_word; i++) {
-            if ((mods[i] + delta) % primes[i] == 0) {
+        for (i = 1; i < NUMPRIMES && primes[i] * primes[i] <= rnd_word; i++) {
+            if (safe ? (mods[i] + delta) % primes[i] <= 1
+                     : (mods[i] + delta) % primes[i] == 0) {
```

Comment:
I think the for loop is now wrong in the case it's not a safe prime

Reply:
No, I don't think so. In the case of not safe primes the loop does a few less iterations.
According to the sieve of Eratosthenes algorithm, you only have to test all primes up to
sqrt(rnd_word).

Reply:
However I need to fix a `warning C4018: '<=': signed/unsigned mismatch` in appveyor here.